src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2125,38 +2127,116 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     regNumber op1Reg;
     regNumber op2Reg;
 
+    regNumber op1NodeReg = op1->GetRegNum();
+    regNumber op2NodeReg = op2->GetRegNum();
+    regNumber op3NodeReg = op3->GetRegNum();
+
+    // For special case that targetReg equals to more than one op Regs
+    bool uniqueOverwrittenReg = true;
+    if ((op1NodeReg == op2NodeReg && targetReg == op1NodeReg) ||
+        (op1NodeReg == op3NodeReg && targetReg == op1NodeReg) || (op2NodeReg == op3NodeReg && targetReg == op2NodeReg))
+        uniqueOverwrittenReg = false;
+
     bool       isCommutative   = false;
     const bool copiesUpperBits = HWIntrinsicInfo::CopiesUpperBits(intrinsicId);
 
     // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
     assert(!copiesUpperBits || !op1->isContained());
 
-    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    // Intrinsics with CopyUpperBits semantics must have op1 as target
+    if ((targetReg == op1NodeReg || copiesUpperBits) && uniqueOverwrittenReg)
     {
-        // 132 form: op1 = (op1 * op3) + [op2]
-
-        ins    = (instruction)(ins - 1);
-        op1Reg = op1->GetRegNum();
-        op2Reg = op3->GetRegNum();
-        op3    = op2;
+        op1Reg = op1NodeReg;
+        if (op2->isContained() || op2->IsRegOptional())
```

Comment:
I believe `IsRegOptional` is only for before register allocation and `isUsedFromSpillTemp` is what you want to represent the same concept in codegen.

IIRC, it basically means `I was reg optional and the register allocator took advantage of that and spilled me".

Reply:
When we set an op be regOptional, is it guaranteed that it will be spilled?

Reply:
No, `IsRegOptional` means that the value can be spilled and codegen can then treat it as contained if it was spilled.

If it doesn't need to be spilled, it won't be.

CC. @kunalspathak

Reply:
I think ```isUsedFromSpillTemp``` means the value is spilled. Therefore,  if ```isUsedFromSpillTemp``` is true, ```IsRegOptional``` is also true. But the opposite is not guaranteed. 

Both ```isUsedFromSpillTemp``` and ```IsRegOptional``` are used in ```hwintrinsiccodegenxarch.cpp``` like here https://github.com/dotnet/runtime/blob/12f260b47fe9e9e89579293b73f1957cc9e94ed0/src/coreclr/jit/hwintrinsiccodegenxarch.cpp#L444.

Reply:
> I think isUsedFromSpillTemp means the value is spilled. Therefore, if isUsedFromSpillTemp is true, IsRegOptional is also true. But the opposite is not guaranteed.

Right, and so the right check here is `if (op2->isContained() || op2->isUsedFromSpillTemp())`. We don't want to force ourselves down this path simply because `op2` could be spilled, only if it actually is spilled.

If the `op2` (the `rhsMul`) is contained or spilled; then `emitOp3` must be `op2` (the `rhsMul`) and we must use the `132` instruction encoding. If `op2` is merely reg optional, but not contained and not spilled, then its already in register and we can technically use `132`, `213`, or `231`; picking based on what the actual target is.